Code cleanup, updated Skript to use lambdas, and some other things

Nicofisi opened the pull request:
In before, by saying "syntax element classes" I mean everything that extends `Effect`, `Condition`, `Expression` et cetera. 

## What I did
I'll try to explain what I did and why as detailed as I can (where there's a need for details)
- Removed the `final` modifier from many `static` methods (not from the parameters). This modifier serves no real purpose, as final static *methods* can be defined in extending classes with the same names regardless.
- Noticed that most syntax element classes used a `final` modifier in parameters of each method (Njol seems to love it lol), but some did not, so I did my best to add them to each method that was missing them.
- Tried to make the syntax element classes use the same order of methods as much as possible. Same thing about the names of parameters, for example I tried to make each `init..` line be exactly this one:
```java
public boolean init(final Expression<?>[] exprs, final int matchedPattern, final Kleenean isDelayed, final ParseResult parseResult) {
```
This really took quite a lot of time..
- Many examples (the Strings in `@Example` annotations) often use indentation at the beginning of some lines. Most of them used the tab character, however I found some that used 4 spaces, and a few with `\t`. I decided to replace each one with \t because it makes it easy to see that tabs were used, and not spaces (the fact that currently the website with docs completely ignores the indentation is another thing).
- Made many package-private fields just private. It causes no harm, plugins couldn't use them anyways and Skript doesn't either so everything's chill.
- In the `init` method the first parameter was usually called `exprs`. In some classes it was called either `vars` or `expressions` or some other names was chosen. I changed the other names to the most popular one - `exprs`
- Fixed some typos in the docs highlighted by IntelliJ, nothing big.
- Some classes had the `javax.annotation.Nullable` imported instead of `org.eclipse.jdt.annotation.Nullable`. In such cases I replaced the import with the Eclipse one.
- Removed some asserts that IntelliJ marked as useless, because they indeed were - they were in some internal classes not accessible by addons and never could evaluate to `false`. 
However, don't worry! I left some unneeded asserts to satisfy the hunger of the #EclipseCompiler
- Removed some of Njol's interesting `String x = "" + something;` patterns. Only some - because actually quite often I could see the reason for using them (the only I could see was to make some strings empty, ones that otherwise would be null)
- Many `private static final class`es had constructors that were `public`. That was pointless so I removed the modifier.
- In a ton of places the diamond operator could be used, so I applied it.
- Some `if`/`else` checks could be simplified without affecting the readability or to be honest even improving it. In such cases I did simplify them with the help of IntelliJ, so you can be sure the logic is right, because the refactoring was performed by the program, not by me.
- Lambdas could be used in many places, shortening the length of code and IMO often shortening the time required to understand some code by a great amount. I wasn't concerned by the support of Java 7 because the project already used lambdas in a few places, so it already wasn't Java-7 compatible. Also if I remember correctly there were every few places where lambdas could be applied but I ignored the suggestion because the code was.. well, way less readable. Yes, I reviewed each case independently.
- Many try blocks had two or more corresponding catch blocks with the same code, I shortened them to `} catch (Exception1 | Exception2 ex) {`
- Added a `//noinspection` comment in many places where `Skript.exception` is called, because this method for some strange reason actually prints the exception in the console, and then returns it, the Exception. Is it even used anywhere? Anyways I added the comment in many places to suppress the warnings from IntelliJ, but after 100+ files I realized how often the method is used and just got humanly bored of suppressing that, sorry 👍 
- Replaced each `Charset.forName("UTF-8")` with `StandardCharsets.UTF_8`
- Added and removed many empty lines, in places where their existence didn't improve the readability at all or there were more than one new lines, one after another, and in places where the code style used in other files screamed "add an empty line to separate these variables please"
- Used IntelliJ to automatically remove all the unneeded imports after I was done with everything else
- Probably some more things that I can't remember at the moment of writing

## License headers
Well, after changing 100+ files I realized the `ctrl + shift + l` shortcut in IntelliJ ("format code nicely") replaced each license header witch this:
```
/**java
 * This file is part of Skript.
 * <p>
 * Skript is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * <p>
 * Skript is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * <p>
 * You should have received a copy of the GNU General Public License
 * along with Skript.  If not, see <http://www.gnu.org/licenses/>.
 * <p>
 * <p>
 * Copyright 2011-2017 Peter Güttinger and contributors
 */
```
If you didn't notice, it added the `<p>` tags. That happened because comments starting with `/**` in Java are supposed to be javadoc comments, and those comments are generally used to generate websites so IntelliJ decided to add some HTML "new paragraph" marks in new lines.

I couldn't find any reason for those license headers to be javadocs. [The license of course doesn't mention it either](https://www.gnu.org/licenses/gpl-3.0.txt):
```
                     END OF TERMS AND CONDITIONS

            How to Apply These Terms to Your New Programs

  If you develop a new program, and you want it to be of the greatest
possible use to the public, the best way to achieve this is to make it
free software which everyone can redistribute and change under these terms.

  To do so, attach the following notices to the program.  It is safest
to attach them to the start of each source file to most effectively
state the exclusion of warranty; and each file should have at least
the "copyright" line and a pointer to where the full notice is found.

    <one line to give the program's name and a brief idea of what it does.>
    Copyright (C) <year>  <name of author>

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
```
Quote: 
> It is safest to attach them to the start of each source file to most effectively
state the exclusion of warranty; and each file should have at least
the "copyright" line and a pointer to where the full notice is found.

So because of lack of good reasons to use javadocs for this and an existence of reasons not to do this I decided to change those headers to just normal comments; I changed each header to this:

```
/*
 * This file is part of Skript.
 * 
 * Skript is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * Skript is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with Skript.  If not, see <http://www.gnu.org/licenses/>.
 * 
 * Copyright 2011-2018 Peter Güttinger and contributors
 */
```

Think optimistically, at least we'll have the year updated!

I really hope you don't mind :)

## What I did not do
- I didn't make any public methods private. I didn't even spend any time thinking on it, I just automatically assumed all public methods and constructors were supposed to be useful either for Skript (via reflection or somehow else) or for addons or for both.
- I didn't touch any warnings that would require actually changing the logic of the code, i.e. I completely ignored (by that I mean *I* ignored them, not that I added a suppress notification, which I didn't) warnings such us `result of mkdirs/renameTo is not used`.
- I didn't touch the `yggdrasil` package, maybe a bit but then thought that it was magic already so noone probably was going to read that anytime soon, saved some time by that ;)
- When seeing some workarounds applied or some commented code that looked like it could be reused in the future I didn't touch that at all. That could make the comments less obvious or not understandable to the reader.

## End words
Of course I'm not perfect and you can probably find a huge amount of places where the things I claim above that I did actually aren't done, but I really tried to be as accurate as possible :) I hope that those changes will make Skript at least a bit easier or cleaner project to work with in general. 

If you disagree with some change please explain with details why. I've spent a ton of time on this PR (well, I was reading a lot of the code at the same time so I suppose I've learnt much about Skript during the process too, and that's a very good thing), so I hope I won't need to revert any change in 650 files, but ugh, I of course am open to some constructive critic, and if there's a good reason to change something, I will try to change it :)

Additionally, I know that those changes are really only the tip of an iceberg, of what needs to be fixed or updated in Skript, but I hope that they will make updating Skript to work with Minecraft 1.13 and working with its source in general at least a bit less painful for our brains :)

Syst3ms commented:
You know you can just change licenseheader.txt to whatever you want and then run `./gradlew licenseFormat`, right ?

Nicofisi commented:
Hmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmm, the more you know XD

bensku commented on build.gradle:
```diff
@@ -1,138 +1,141 @@
+        maven {
+            url "file:///itemapi_repo/"
+        }
+    }
```
Skript's build files use tabs, and that is intentional.

Nicofisi replied:
Fine, gonna rollback, IntelliJ changed that for me

Nicofisi replied:
"That is intentional" any reason for that actually?

bensku commented on licenseheader.txt:
```diff
@@ -1,17 +1,16 @@
-  This file is part of Skript.
+This file is part of Skript.
```
While having spaces before license is not stricly necessary, there is also no reason to change it. Or is there?

bensku replied:
Well, it causes a mess while looking at diffs. More lines changed than necessary. I'd almost think the license header contents were changed, not just first line (no Javadoc) and year of copyright.

But no, it does not really matter that much. People don't read diffs of license headers.

bensku commented on src/main/java/ch/njol/skript/Metrics.java:
```diff
@@ -28,1010 +29,1013 @@
 /**
  * bStats collects some data for plugin authors.
- *
+ * <p>
```
Comes from Bstats, do not edit this file unless absolutely necessary.

Nicofisi replied:
Sure, gonna rollback too

bensku commented on src/main/java/ch/njol/skript/ScriptLoader.java:
```diff
@@ -93,45 +63,68 @@
 final public class ScriptLoader {
-	private ScriptLoader() {}
-	
+	private ScriptLoader() {
```
This was perfectly fine.

Nicofisi replied:
This wasn't on purpose, I made IntelliJ not change lines like this after changing about 50 files already, let me know if you see this expanded anywhere else please

bensku commented on src/main/java/ch/njol/skript/ScriptLoader.java:
```diff
@@ -93,45 +63,68 @@
 			m_no_scripts = new Message("skript.no scripts");
 	private final static PluralizingArgsMessage m_scripts_loaded = new PluralizingArgsMessage("skript.scripts loaded");
-	
+
```
Seriously, no need to remove tabs in empty lines. Doesn't matter for readability.

Nicofisi replied:
Ugh, this would be a big pain to try to revert that, is this change really very annoying? It affects literally every file

bensku replied:
That is why it is annoying; it affects all files. Before merging, someone will have to read through the whole set of changed files. Which means about 650 files... It is going to take me or @Snow-Pyon ages.

But yeah, we can do it. Will just take _some_ time :)

bensku commented on src/main/java/ch/njol/skript/ScriptLoader.java:
```diff
@@ -192,34 +185,34 @@ public void add(final ScriptInfo other) {
 	// Load scripts in separate (one) thread
 	static final BlockingQueue<Runnable> loadQueue = new ArrayBlockingQueue<>(20, true);
-	static final Thread loaderThread;
+	private static final Thread loaderThread;
```
Asynchronous loading is not entirely finished. This might be needed from elsewhere.

bensku commented on src/main/java/ch/njol/skript/ScriptLoader.java:
```diff
@@ -230,114 +223,109 @@ public String toString() {
 				try {
 					loadQueue.take().run();
 				} catch (InterruptedException e) {
+					//noinspection ThrowableNotThrown
```
IDE-specific warning suppression comments are not necessary. Please just configure your IDE to not create the warnings, if they are not for a reason. But chances are that the warnings _do_ have a reason, in which case that should be fixed.

bensku replied:
I don't think there is any reason. However, changing signatures of public methods can produce endless amounts of _fun_ when people use addons...

Nicofisi replied:
Oh yeah right, I guess it's one of the top-used methods by addons (in the init methods at least), didn't think about that

bensku commented on src/main/java/ch/njol/skript/ScriptLoader.java:
```diff
@@ -230,114 +223,109 @@ public String toString() {
 	 * @return Info on the loaded scripts.
 	 */
-	public final static ScriptInfo loadScripts(final List<Config> configs) {
+	public static ScriptInfo loadScripts(final List<Config> configs) {
```
https://stackoverflow.com/questions/1743715/behaviour-of-final-static-method

Final modifier with static methods has its purpose. I'm not sure if it is important one, though. Opinions?

bensku commented on src/main/java/ch/njol/skript/ScriptLoader.java:
```diff
@@ -380,95 +368,95 @@ public final static ScriptInfo loadScripts(final List<Config> configs, final Lis
 	private static class ParsedEventData {
-		
-		public ParsedEventData(NonNullPair<SkriptEventInfo<?>, SkriptEvent> info, String event, SectionNode node, List<TriggerItem> items) {
+		ParsedEventData(NonNullPair<SkriptEventInfo<?>, SkriptEvent> info, String event, SectionNode node, List<TriggerItem> items) {
```
A class, no matter if inner class, should have first line free. Improves readibility.

bensku commented on src/main/java/ch/njol/skript/ScriptLoader.java:
```diff
@@ -500,25 +490,23 @@ private final static ScriptInfo loadScript(final @Nullable Config config) {
 							}
-							String name = ((EntryNode) n).getKey().toLowerCase(Locale.ENGLISH);
+							String key = n.getKey();
+							assert key != null; // Eclipse
```
Wait a second... Are you sure that the key cannot be null? I think getKey has been explicitly marked as nullable?

Syst3ms replied:
getKey is nullable in Node, but non-null in most extending classes, including EntryNode

bensku replied:
Ah yeah. EntryNode check is somewhere above this diff piece, right? Seems good, then.

bensku commented on src/main/java/ch/njol/skript/ScriptLoader.java:
```diff
@@ -736,86 +720,92 @@ public String run(final Matcher m) {
+	 *
 	 * @param source Source input stream.
-	 * @param name Name of source "file".
+	 * @param name   Name of source "file".
```
One space after parameter name.

Nicofisi replied:
IntelliJ tries to align this to columns automatically. Do you think it's a bad practice? I wouldn't say so, Java itself does that, look for example here http://www.docjar.net/html/api/java/util/Collections.java.html
```
   62    * @author  Josh Bloch
   63    * @author  Neal Gafter
   64    * @see     Collection
   65    * @see     Set
   66    * @see     List
   67    * @see     Map
   68    * @since   1.2
```
I'm pretty sure Eclipse should have some option for this too

Nicofisi replied:
I mean, javadocs are theoretically meant to probably mainly be used with a website, but I personally (and I don't think I stand out much here) 98% of the time just look at them along with the code in the IDE, either source or decompiled, and this improves readability greatly, and is automated, so I'd say it's good

bensku commented on src/main/java/ch/njol/skript/SkriptConfig.java:
```diff
@@ -50,230 +34,185 @@
-	static Collection<Config> configs = new ArrayList<Config>();
-	
-	final static Option<String> version = new Option<String>("version", Skript.getVersion().toString())
+	private static Config mainConfig;
```
Probably intentional to have it package-private. Or is there getter somewhere?

Nicofisi replied:
Reverted

bensku commented on src/main/java/ch/njol/skript/SkriptConfig.java:
```diff
@@ -50,230 +34,185 @@
-					SkriptLogger.setVerbosity(v);
+
+	final static Option<Verbosity> verbosity = new Option<>("verbosity", Verbosity.NORMAL, new EnumParser<>(Verbosity.class, "verbosity"))
+			.setter(SkriptLogger::setVerbosity);
```
Are you sure this is functionally equivalent?

bensku replied:
Internally, method handles as lambdas are quite a nice piece of spaghetti and runtime code generation. But if you're sure it works, fine.

bensku commented on src/main/java/ch/njol/skript/SkriptConfig.java:
```diff
@@ -50,230 +34,185 @@
-				
-			})
+	public final static Option<Boolean> colorResetCodes = new Option<>("color codes reset formatting", true)
+			.setter(str -> ChatMessages.colorResetCodes = str);
```
str? It isn't a string

Nicofisi replied:
My bad, fixed

bensku commented on src/main/java/ch/njol/skript/SkriptEventHandler.java:
```diff
@@ -190,27 +185,25 @@ public static void logTriggerEnd(final Trigger t) {
-			List<Trigger> ts = triggers.get(e);
-			if (ts == null)
-				triggers.put(e, ts = new ArrayList<>());
+			List<Trigger> ts = triggers.computeIfAbsent(e, k -> new ArrayList<>());
```
Lambdas cause object creation, and the object might become garbage if stack allocation does not get it. This is to be avoided in functions that are called (very) often.

Syst3ms replied:
Why doesn't this use Guava's Multimap anyway ?

bensku replied:
Up-to-date Guava has been exposed to plugins only for a few years. And besides, this might be faster. Benchmark before changing this, I guess...

Syst3ms replied:
iirc Multimap is a quite old Guava class, let me check that.

Nicofisi replied:
What about that, should I revert that to how it was? `triggers.computeIfAbsent` occur in SkriptEventHandler and EvtMoveOn now

bensku commented on src/main/java/ch/njol/skript/Updater.java:
```diff
@@ -98,79 +85,71 @@
 	 */
 	@NonNullByDefault(value = false)
 	public class ResponseEntry {
 		public String url;
```
Good, replaced spaces with tabs. That is correct.

bensku commented on src/main/java/ch/njol/skript/Updater.java:
```diff
@@ -181,30 +160,26 @@ public static void start() {
 		assert str != null : "Cannot deserialize null string";
 		@SuppressWarnings("serial")
-		Type listType = new TypeToken<List<ResponseEntry>>() {}.getType();
+		Type listType = new TypeToken<List<ResponseEntry>>() {
```
This is meant to be one-liner.

bensku commented on src/main/java/ch/njol/skript/Updater.java:
```diff
@@ -344,7 +312,8 @@ public DownloaderTask(Plugin plugin) {
 		private CommandSender sender;
-		
+
+		@SuppressWarnings("ThrowableNotThrown")
```
Is this standard Javac warning suppressed? If not, it is unnecessary. Only nullness analysis and $CASES-OMITTED$ get pass here; they are quite absolutely necessary in Skript.

Nicofisi replied:
What is cases ommited used by btw? The compiler or some debug tool?

bensku replied:
It is a comment that tells some IDEs and maybe Javac that an enum switch does not handle all cases.

Nicofisi replied:
Hmm, are you sure this isn't only for the Eclipse IDE? I mean, fine if it is I suppose, but just wondering, I can't find anything on Google

bensku commented on src/main/java/ch/njol/skript/aliases/ItemData.java:
```diff
@@ -178,68 +176,64 @@ public ItemStack getRandom() {
 	}
-	
+
+	@SuppressWarnings("MethodDoesntCallSuperMethod")
```
Not standard warning. Not needed.

Nicofisi replied:
Ok I guess that warning I'll just disable for the whole project, I don't see how this warning can really useful in places, excluding than maybe Android projects tbf (unless clone() really should always call the super method and Njol used it wrong)

bensku replied:
Clone does not need to call super.clone() as long as it produces a copy. What kind of copy, is not defined; if it is not a deep copy, Javadoc should probably mention it. But even that is optional.

bensku commented on src/main/java/ch/njol/skript/aliases/ItemType.java:
```diff
@@ -1,49 +1,23 @@
-/**
- *   This file is part of Skript.
+/*
```
Aliases rework reworks this file quite completely.

bensku commented on src/main/java/ch/njol/skript/aliases/MaterialName.java:
```diff
@@ -1,67 +1,66 @@
-/**
- *   This file is part of Skript.
+/*
```
Aliases rework here too.

Nicofisi replied:
What about that, should I revert any changes? I guess it still can be merged easily

bensku replied:
This will cause a merge conflict, which I can solve in about 5 seconds. The point was, don't spend time on these classes even if you decide to do more refactorings in other places. It would not be useful.

bensku commented on src/main/java/ch/njol/skript/bukkitutil/BiomeMappings.java:
```diff
@@ -1,58 +1,58 @@
-	public static @Nullable Biome parse(final String name) {
+
+	public static @Nullable
+	Biome parse(final String name) {
```
Two lines? Umm, does not seem like a good idea.

Nicofisi replied:
My fault, sorry. I don't think so either

Nicofisi replied:
Actually it's the fault of how it was written that IntelliJ chose to place it this way, the @Nullable in every other place is in the line above the `modifiers ReturnType name(args) {`

bensku replied:
Oh, that explains a lot. I wonder what I was thinking when I wrote the annotation that way.

bensku commented on src/main/java/ch/njol/skript/bukkitutil/HealthUtils.java:
```diff
@@ -51,103 +51,100 @@ private HealthUtils() {}
 			return ((Number) getHealth.invoke(e)).doubleValue() / 2;
-		} catch (final IllegalAccessException ex) {
+		} catch (final IllegalAccessException | InvocationTargetException ex) {
+			//noinspection ThrowableNotThrown
```
???

bensku replied:
Ah yeah. Same thing as above, then.

bensku commented on src/main/java/ch/njol/skript/bukkitutil/PlayerUtils.java:
```diff
@@ -27,32 +35,22 @@
+ *
  * @author Peter Güttinger
  */
+@SuppressWarnings({"ThrowableNotThrown", "FinalStaticMethod"})
```
Nonstandard warnings suppressed. And besides, final static methods are a warning... ?

bensku replied:
Ah, I see.

Syst3ms commented:
> x = "" + something? Could you give an example, or should I just dig through all the changes?

https://github.com/bensku/Skript/blob/master/src/main/java/ch/njol/skript/lang/SkriptParser.java#L112

SkriptParser is cluttered with these abominations, as if it weren't enough of an abomination already.

Nicofisi commented:
Actually I see I missed most of those useless `"" +`s now, I didn't read every single line of code in Skript during the refactoring, hence why xD

Syst3ms commented:
Also, `final` on `static` methods only makes a noticeable difference when static methods are called from instances, which is an awful thing to do, so I can safely say that's useless.

bensku commented:
"" + something is always non-null. I wonder if some ancient ECJ version did not know that trimming a string can never return null...

Syst3ms commented:
I doubt it thought substringing could return null, though.

Nicofisi commented:
Hmm, @bensku do you really think those empty lines add much of readability? To be fair removing those lines at the beginning and end of classes was probably the first thing I did manually after opening each file.. 

A few times when I wasn't sure about how something should be formatted I looked at Google's Java Syntax Guide, [where there's written](https://google.github.io/styleguide/javaguide.html#s4.6.1-vertical-whitespace):
> [Vertical Whitespace]
> [A single blank line appears:]
> *Optionally* before the first member or initializer, or after the last member or initializer of the class (neither encouraged nor discouraged).

To be fair to me, at least on the ends, it always looked like the class was incomplete and someone was going to add something there and forgot or changed their mind, hmmm. Also for example, if we don't do this:
```java
if (something) {
    // code
    if (somethingelse) {
        // code
    }

}
// I'm talking about the empty line two lines above
```
is there a good reason to do that after the end of the last method, where the ending is the same? 

If you really consider this as something that improves the readability significantly I can try to perhaps find an option for that in the IDE code style settings and right click the whole project and click refactor, but I believe that could cause many unexpected refactoring in other parts of the classes. For the rest of the PR I've always been refactoring one file at a time and doing my best while scrolling through the file to confirm that nothing was changed weirdly (sometimes it was), just in case. The only other option I know would be to change that manually in each file separately, which would require 1/2 hours of fast manual editing with no thinking involved. Any thoughts/tool suggestions? Rip

xXAndrew28Xx commented:
Maybe you should move bStats from a copied class to a shadowed in jar.

Nicofisi commented:
@xXAndrew28Xx you mean to just delete the class and compile it as a lib from their repo in gradle?

xXAndrew28Xx commented:
Yeah, have it as a dependency and shadow it in.

bensku commented:
@xXAndrew28Xx  Shading is out of scope for this PR. I set up Gradle shadow plugin in another branch, but it has just not been merged to master yet.

bensku commented:
@Nicofisi Learnt to like empty lines at start of classes a few years ago when working with SpongeAPI. I'm not sure why it feels more readable to me, though. Probably a bit more clear where class declaration ends and where contents begin.

At ends, I don't intentionally add empty lines. Because a class ends with just } (no fancy and long generics declarations here), it is completely unnecessary. I mean, sometimes when a class is indeed work in progress, that might happen.

Empty lines at ends of methods and statements are also usually not intentional and useless. Separating two conditional blocks with an empty line is useful; that empty line should be between them, not inside either of them. For methods, it is common to have empty lines between them in class, so same thing applies.

Syst3ms commented:
What I personally do is have one empty line when I only have a constructor/methods, and none of I have fields.

bensku commented:
Oops. Was fixing fallout from another PR (docs additions) and added some Javadoc too. Sadly it led to a merge conflict. But I guess we can just take master version for docs files, because they are now quite clean.

Nicofisi commented:
* Fixed the conflicts
* Removed IntelliJ-specific warning suppress annotations pointed out by @bensku 
* Fixed some places where `{}` was changed to `{\n<spaces>}`
* `build.gradle` uses tabs again
* Any changes to `Metrics.java` are reverted
* *Maybe something else is reverted*
* *Maybe something else but I don't think so*

Nicofisi commented on src/main/java/ch/njol/yggdrasil/DefaultYggdrasilOutputStream.java:
```diff
@@ -196,8 +196,7 @@ protected void writeArrayLength(final int length) throws IOException {
 	@Override
-	protected void writeArrayEnd() throws IOException {
-	}
+	protected void writeArrayEnd() throws IOException {}
```
Actually this probably should even before be
```java
method() {
  // empty line here
}
```
shouldn't it? or is it fine

bensku commented:
Ok, probably time to go through all the changed files. Will take a few days, probably. Please do not do any more changes unless requested - they would cause additional delay.

Seems good so far.

bensku commented:
Will probably take a bit longer. Have been busy lately.

Nicofisi commented:
Alright fine

bensku commented:
I have review in progress. At least 500 files remaining, so it'll take a "while".

Nicofisi commented:
It won't compile right now, I need to put asserts back in two places, other than that should be "fine"

Syst3ms commented:
@Nicofisi Travis CI is complaining because the `string` field is nullable in VariableString, yet you're doing `string[i]`.

Nicofisi commented:
I know right i just committed without checking and couldn't find the time to fix it

bensku commented:
@Nicofisi You reversed import order and added a few * imports. These things are not to be done in Skript, and unfortunately there is no solution but revert said changes. Also, while optional for subclasses, I want root class of each file to have an empty line as first line inside the class.

To be honest... I've read through like 100 files, which have only changes to license headers or empty lines where spaces were removed. Both of those changes are kinda out of scope of improving code quality, too.

But fear not! You didn't throw everything to a single commit, so cherry picking individual changes should be possible. If nothing else helpls, that is.

bensku commented:
Merged one commit, apparently causing a conflict. Fixing right now.

bensku commented:
So, merged @TheLimeGlass's huge PR. It added quite a few new features and such, so decided to prioritize it over this.

My recommendation: start a new pull request and (cherry-)pick changes from there that seem good. Alternatively, I think I can do some of that for you if you want.

Nicofisi commented:
See #1314
